presto-main/src/main/java/com/facebook/presto/util/JsonUtil.java

```diff
@@ -38,4 +61,680 @@ public static JsonGenerator createJsonGenerator(JsonFactory factory, SliceOutput
     {
         return factory.createGenerator((OutputStream) output);
     }
+
+    public static boolean canCastToJson(Type type)
+    {
+        String baseType = type.getTypeSignature().getBase();
+        if (type instanceof UnknownType ||
+                baseType.equals(StandardTypes.BOOLEAN) ||
+                baseType.equals(StandardTypes.TINYINT) ||
+                baseType.equals(StandardTypes.SMALLINT) ||
+                baseType.equals(StandardTypes.INTEGER) ||
+                baseType.equals(StandardTypes.BIGINT) ||
+                baseType.equals(StandardTypes.REAL) ||
+                baseType.equals(StandardTypes.DOUBLE) ||
+                baseType.equals(StandardTypes.DECIMAL) ||
+                baseType.equals(StandardTypes.VARCHAR) ||
+                baseType.equals(StandardTypes.JSON)) {
+            return true;
+        }
+        if (type instanceof ArrayType) {
+            return canCastToJson(((ArrayType) type).getElementType());
+        }
+        if (type instanceof MapType) {
+            return isValidJsonObjectKeyType(((MapType) type).getKeyType()) && canCastToJson(((MapType) type).getValueType());
+        }
+        if (type instanceof RowType) {
+            return type.getTypeParameters().stream().allMatch(fieldType -> canCastToJson(fieldType));
+        }
+        return false;
+    }
+
+    private static boolean isValidJsonObjectKeyType(Type type)
+    {
+        String baseType = type.getTypeSignature().getBase();
+        return type instanceof UnknownType ||
+                baseType.equals(StandardTypes.BOOLEAN) ||
+                baseType.equals(StandardTypes.TINYINT) ||
+                baseType.equals(StandardTypes.SMALLINT) ||
+                baseType.equals(StandardTypes.INTEGER) ||
+                baseType.equals(StandardTypes.BIGINT) ||
+                baseType.equals(StandardTypes.REAL) ||
+                baseType.equals(StandardTypes.DOUBLE) ||
+                baseType.equals(StandardTypes.DECIMAL) ||
+                baseType.equals(StandardTypes.VARCHAR);
+    }
+
+    // Read the map key as the Json object key format (string)
+    public abstract static class MapObjectKeyReader
```

Comment:
Make this an interface, and use lambda when constructing implementations of this interface